src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,462 @@
+{
+    if (!e.exitIsExit() || e.outIsEmpty()) {
+        return;
+    }
+    stream.writeStartElement("exit");
+    stream.writeAttribute("dir", lowercaseDirection(dir));
+    saveExitTo(stream, e);
+    saveXmlAttribute(stream, "doorname", e.getDoorName().toQString());
+    saveDoorFlags(stream, e.getDoorFlags());
+    saveExitFlags(stream, e.getExitFlags());
+    stream.writeEndElement(); // end exit
+}
+
+void XmlMapStorage::saveExitTo(QXmlStreamWriter &stream, const Exit &e)
+{
+    QString idlist;
+    const char *separator = nullptr;
+    for (const RoomId id : e.outRange()) {
+        if (separator == nullptr) {
+            separator = ",";
+        } else {
+            idlist += separator;
+        }
+        idlist += QString("%1").arg(id.asUint32());
+    }
+    stream.writeAttribute("to", idlist);
+}
+
+void XmlMapStorage::saveMarkers(QXmlStreamWriter &stream, const MarkerList &markerList)
+{
+    ProgressCounter &progressCounter = getProgressCounter();
+    for (const auto &marker : markerList) {
+        saveMarker(stream, deref(marker));
+        progressCounter.step();
+    }
+}
+
+void XmlMapStorage::saveMarker(QXmlStreamWriter &stream, const InfoMark &marker)
+{
+    const InfoMarkTypeEnum type = marker.getType();
+    stream.writeStartElement("marker");
+    stream.writeAttribute("type", markTypeName(type));
+    stream.writeAttribute("class", markClassName(marker.getClass()));
+    // REVISIT: round to 45 degrees?
+    if (marker.getRotationAngle() != 0) {
+        stream.writeAttribute("angle",
+                              QString("%1").arg(static_cast<qint32>(marker.getRotationAngle())));
+    }
+    saveCoordinate(stream, "pos1", marker.getPosition1());
+    saveCoordinate(stream, "pos2", marker.getPosition2());
+
+    if (type == InfoMarkTypeEnum::TEXT) {
+        stream.writeStartElement("text");
+        stream.writeCharacters(marker.getText().toQString());
+        stream.writeEndElement(); // end text
+    }
+
+    stream.writeEndElement(); // end marker
+}
+
+void XmlMapStorage::saveXmlElement(QXmlStreamWriter &stream,
+                                   const QString &name,
+                                   const QString &value)
+{
+    if (!value.isEmpty()) {
+        stream.writeStartElement(name);
+        stream.writeCharacters(value);
+        stream.writeEndElement(); // end description
+    }
+}
+
+void XmlMapStorage::saveXmlAttribute(QXmlStreamWriter &stream,
+                                     const QString &name,
+                                     const QString &value)
+{
+    if (!value.isEmpty()) {
+        stream.writeAttribute(name, value);
+    }
+}
+
+void XmlMapStorage::log(const QString &msg)
+{
+    emit sig_log("XmlMapStorage", msg);
+}
+
+void XmlMapStorage::saveDoorFlags(QXmlStreamWriter &stream, DoorFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    for (uint x = 0; x < NUM_DOOR_FLAGS; x++) {
```

Comment:
The X_FOREACH_DOOR_FLAG macro would work here too

Reply:
same considerations as previous point above

Reply:
done too, see previous point above